pkg/k8s/watchers/cilium_endpoint.go

```diff
@@ -116,13 +116,13 @@ func (k *K8sWatcher) endpointUpdated(oldEndpoint, endpoint *types.CiliumEndpoint
 					}
 				}
 				if !v4Added {
-					portsChanged := k.ipcache.DeleteOnMetadataMatch(oldPair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name)
+					portsChanged := k.ipcache.DeleteOnMetadataMatch(oldPair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name, string(endpoint.UID))
```

Comment:
Good point, I'll fix this up.

Reply:
<details><summary>Details</summary>
<p>

```diff
commit b6be38f8e5de748c852376f87a2737d09a83855d
Author: Chris Tarazi <chris@isovalent.com>
Date:   Tue Apr 9 14:02:34 2024 -0700

    fixup! ipcache,watchers,identity: Pass pod UID into ipcache metadata
    
    Signed-off-by: Chris Tarazi <chris@isovalent.com>

diff --git a/api/v1/models/ip_list_entry_metadata.go b/api/v1/models/ip_list_entry_metadata.go
index 75a0099252..6efafe32e9 100644
--- a/api/v1/models/ip_list_entry_metadata.go
+++ b/api/v1/models/ip_list_entry_metadata.go
@@ -29,6 +29,9 @@ type IPListEntryMetadata struct {
 	// Source of the IP entry and its metadata
 	// Example: k8s
 	Source string `json:"source,omitempty"`
+
+	// UID of the backing resource to the IP (e.g. Kubernetes pod UID)
+	UID string `json:"uid,omitempty"`
 }
 
 // Validate validates this IP list entry metadata
diff --git a/api/v1/openapi.yaml b/api/v1/openapi.yaml
index 3a773d5ceb..ca93a92e1b 100644
--- a/api/v1/openapi.yaml
+++ b/api/v1/openapi.yaml
@@ -3787,6 +3787,9 @@ definitions:
       name:
         description: Name assigned to the IP (e.g. Kubernetes pod name)
         type: string
+      uid:
+        description: UID of the backing resource to the IP (e.g. Kubernetes pod UID)
+        type: string
   NamedPorts:
     description: |-
       List of named Layer 4 port and protocol pairs which will be used in Network
diff --git a/api/v1/server/embedded_spec.go b/api/v1/server/embedded_spec.go
index 7314db4792..37e3b02204 100644
--- a/api/v1/server/embedded_spec.go
+++ b/api/v1/server/embedded_spec.go
@@ -3573,6 +3573,10 @@ func init() {
           "description": "Source of the IP entry and its metadata",
           "type": "string",
           "example": "k8s"
+        },
+        "uid": {
+          "description": "UID of the backing resource to the IP (e.g. Kubernetes pod UID)",
+          "type": "string"
         }
       }
     },
@@ -9423,6 +9427,10 @@ func init() {
           "description": "Source of the IP entry and its metadata",
           "type": "string",
           "example": "k8s"
+        },
+        "uid": {
+          "description": "UID of the backing resource to the IP (e.g. Kubernetes pod UID)",
+          "type": "string"
         }
       }
     },
diff --git a/daemon/cmd/ipcache.go b/daemon/cmd/ipcache.go
index 31d85eac9a..9fdf84792a 100644
--- a/daemon/cmd/ipcache.go
+++ b/daemon/cmd/ipcache.go
@@ -86,6 +86,7 @@ func (ipc *ipCacheDumpListener) OnIPIdentityCacheChange(modType ipcache.CacheMod
 			Source:    string(newID.Source),
 			Namespace: k8sMeta.Namespace,
 			Name:      k8sMeta.PodName,
+			UID:       k8sMeta.UID,
 			// TODO (jrajahalme): Consider if named ports should be
 			//                    made visible in the model.
 		}
diff --git a/pkg/k8s/factory_functions.go b/pkg/k8s/factory_functions.go
index 46eec43c94..42ef3a411f 100644
--- a/pkg/k8s/factory_functions.go
+++ b/pkg/k8s/factory_functions.go
@@ -457,8 +457,9 @@ func TransformToCiliumEndpoint(obj interface{}) (interface{}, error) {
 				ResourceVersion: concreteObj.ObjectMeta.ResourceVersion,
 				// We don't need to store labels nor annotations because
 				// they are not used by the CEP handlers.
-				Labels:      nil,
-				Annotations: nil,
+				Labels:          nil,
+				Annotations:     nil,
+				OwnerReferences: slim_metav1.SlimOwnerReferences(concreteObj.OwnerReferences),
 			},
 			Encryption: func() *cilium_v2.EncryptionSpec {
 				enc := concreteObj.Status.Encryption
@@ -492,8 +493,9 @@ func TransformToCiliumEndpoint(obj interface{}) (interface{}, error) {
 					ResourceVersion: ciliumEndpoint.ObjectMeta.ResourceVersion,
 					// We don't need to store labels nor annotations because
 					// they are not used by the CEP handlers.
-					Labels:      nil,
-					Annotations: nil,
+					Labels:          nil,
+					Annotations:     nil,
+					OwnerReferences: slim_metav1.SlimOwnerReferences(ciliumEndpoint.ObjectMeta.OwnerReferences),
 				},
 				Encryption: func() *cilium_v2.EncryptionSpec {
 					enc := ciliumEndpoint.Status.Encryption
diff --git a/pkg/k8s/watchers/cilium_endpoint.go b/pkg/k8s/watchers/cilium_endpoint.go
index 5e62d3273e..7cc44c1288 100644
--- a/pkg/k8s/watchers/cilium_endpoint.go
+++ b/pkg/k8s/watchers/cilium_endpoint.go
@@ -167,10 +167,11 @@ func (k *K8sWatcher) endpointUpdated(oldEndpoint, endpoint *types.CiliumEndpoint
 		return
 	}
 
+	podUID := k.fetchPodUIDForCEP(endpoint)
 	k8sMeta := &ipcache.K8sMetadata{
 		Namespace:  endpoint.Namespace,
 		PodName:    endpoint.Name,
-		UID:        string(endpoint.UID),
+		UID:        podUID,
 		NamedPorts: make(ciliumTypes.NamedPortMap, len(endpoint.NamedPorts)),
 	}
 	for _, port := range endpoint.NamedPorts {
@@ -206,18 +207,19 @@ func (k *K8sWatcher) endpointUpdated(oldEndpoint, endpoint *types.CiliumEndpoint
 }
 
 func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
+	podUID := k.fetchPodUIDForCEP(endpoint)
 	if endpoint.Networking != nil {
 		namedPortsChanged := false
 		for _, pair := range endpoint.Networking.Addressing {
 			if pair.IPV4 != "" {
-				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name, string(endpoint.UID))
+				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name, podUID)
 				if portsChanged {
 					namedPortsChanged = true
 				}
 			}
 
 			if pair.IPV6 != "" {
-				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV6, source.CustomResource, endpoint.Namespace, endpoint.Name, string(endpoint.UID))
+				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV6, source.CustomResource, endpoint.Namespace, endpoint.Name, podUID)
 				if portsChanged {
 					namedPortsChanged = true
 				}
@@ -228,3 +230,20 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
 		}
 	}
 }
+
+func (k *K8sWatcher) fetchPodUIDForCEP(endpoint *types.CiliumEndpoint) string {
+	// Fetch the Pod UID from the internal Endpoint object if it exists,
+	// otherwise check the ownerRef of the CEP object.
+	var podUID string
+	ep := k.endpointManager.LookupCEPName(endpoint.Namespace + "/" + endpoint.Name)
+	if ep != nil {
+		// If the endpoint is local to this node, then this should suffice.
+		podUID = ep.GetK8sUID()
+	}
+	if podUID == "" {
+		if len(endpoint.OwnerReferences) > 0 && endpoint.OwnerReferences[0].Kind == "Pod" {
+			podUID = string(endpoint.OwnerReferences[0].UID)
+		}
+	}
+	return podUID
+}
```

</p>
</details>

Reply:
I'm not sure I understand what else `endpointUpdate` needs?